[FIX] contract: multi-company assignment notification

yajo opened the pull request:
Steps to reproduce the problem:

1. Log in as Mitchell Admin.
2. Create contract CNT-A for company CMP-A, assigned to Marc Demo.
3. Create contract CNT-B for company CMP-B, assigned to Marc Demo.
4. Run cron to create recurring invoices.

Actual results:

- Odoo sends automated assignment emails to Marc Demo, which indicate the name of the company activated for `__system__` user whlie the cron was being executed.

Expected results after this patch:

- Odoo sends automated assignment emails to Marc Demo, which indicate the invoice company.

@Tecnativa TT24657

pedrobaeza commented:
I don't really like to mangle with the user's company, as there can be other operations in parallel that are breaking due to this change. Have you checked how that assignation notification is done instead? I prefer to modify that code to take document's company instead.

yajo commented:
I don't think that is really going to happen because there's no DB commit between the company changes, so theoretically the company is never going to change in other cursors. I don't know if odoo will end up issuing an `UPDATE` query, but in any case it's going to update to the same value, so it shouldn't clash.

Before you ask, the notification mails are queued, not sent directly, so they don't produce a DB commit or race condition either.

> Have you checked how that assignation notification is done instead?

Considering what I mentioned above, I didn't see this necessary; but yes, I have checked this option and it would mean to patch [this line](https://github.com/odoo/odoo/blob/bfd2b669c056cc9fa6e97992944e66aac7c1092a/odoo/addons/base/models/ir_ui_view.py#L1343).

Do you still think the change is needed/benefitial? :thinking: 

---

BTW neither of these will be needed in v13, where all can be fixed with a context key.

pedrobaeza commented:
Well, we don't know what overrides are in the system, and modifying wildly that company for whole process seems to risky for me, and more knowing that this is not a problem anymore in v13, so I would write this in the ROADMAP of v12 and notify users that it won't be a problem there (but are you sure the link to Odoo reference is changed in v13 to take document's company?). I would prefer the patch in Odoo to take there the document's company.

yajo commented:
> are you sure the link to Odoo reference is changed in v13 to take document's company?

Nope, it takes the company from the context. We will still need the workaround, it will just be simpler.

> I would prefer the patch in Odoo to take there the document's company.

That is a big patch, and it doesn't affect upstream code, so I see too many possibilities that Odoo rejects it.

Let me just write the alternative workaround here and add the roadmap note.

yajo commented:
Patch ready.

pedrobaeza reviewed (approved):
OK, it can work in a less risky way.

sbejaoui commented on contract/models/contract.py:
```diff
@@ -546,8 +546,13 @@ def cron_recurring_create_invoice(self, date_ref=None):
+            contracts_to_invoice = self.search(row["__domain"]).with_context(
+                allowed_company_ids=[row["company_id"][0]]
+            )
+            contracts_to_invoice._recurring_create_invoice(date_ref)
```
please keep the return

pedrobaeza replied:
invoices records should be then stacked and return at the end.

sbejaoui replied:
yes that what I mean :)

yajo replied:
I saw it's used nowhere, but OK no problem on this.

sbejaoui reviewed (changes requested):
We didn't notice this case as we configure one cron per company with a regular user.

Commit:
[FIX] contract: multi-company assignment notification

Steps to reproduce the problem:

1. Log in as Mitchell Admin.
2. Create contract CNT-A for company CMP-A, assigned to Marc Demo.
3. Create contract CNT-B for company CMP-B, assigned to Marc Demo.
4. Run cron to create recurring invoices.

Actual results:

- Odoo sends automated assignment emails to Marc Demo, which indicate the name of the company activated for `__system__` user whlie the cron was being executed.

Expected results after this patch:

- Odoo sends automated assignment emails to Marc Demo, which indicate the invoice company.

@Tecnativa TT24657

yajo commented:
All fixed.

sbejaoui reviewed (approved):
thanks @Yajo

pedrobaeza commented:
/ocabot merge patch

pedrobaeza commented:
Included this in v13 in latest push:

https://github.com/OCA/contract/compare/21ff7d808df330a728fe56dff401a6da4ee65a77..647a4417466e6264fc1c36976992735d7ec247b2

yajo commented:
Mm? What do you mean? The addon isn't in v13 yet AFAICS.

pedrobaeza commented:
I added to the migration PR: https://github.com/OCA/contract/pull/533

You can review it by chance :angel:
